packages/block-editor/src/components/media-placeholder/index.js

```diff
@@ -116,14 +116,38 @@ export class MediaPlaceholder extends Component {
 			multiple,
 			onError,
 			onSelect,
-			value = [],
 		} = this.props;
 		let setMedia;
 		if ( multiple ) {
 			if ( addToGallery ) {
-				const currentValue = value;
+				// To allow changes to a gallery to be made while uploads are in progress
+				// (including trigging multiple upload groups and removing already in place images),
+				// we must be able to add newMedia based on the current value of the Gallery
+				// whenever the setMedia function runs. Since the setMedia function runs multiple
+				// times per upload group and is passed newMedia containing every item in its
+				// group each time, we must also filter out whatever this upload group had
+				// previously returned to the gallery before adding and returning the image
+				// array with replacement newMedia values.
+
+				// Define an array to store urls from newMedia between subsequent function calls.
+				let lastMediaPassed = [];
 				setMedia = ( newMedia ) => {
-					onSelect( currentValue.concat( newMedia ) );
+					// Get the current array of images on the Gallery.
+					const currentMedia = this.props.value || [];
+					// Remove any images this upload group is responsible for (lastMediaPassed).
+					// Their replacements are contained in newMedia.
+					const filteredMedia = currentMedia.filter( ( item ) => {
+						// Compare via .includes since gallery adds a ratio to the url after returned from this function.
+						return ! lastMediaPassed.some( ( temporaryMediaSlug ) => item.url.includes( temporaryMediaSlug ) );
+					} );
+					// Return the filtered media array along with newMedia.
+					onSelect( filteredMedia.concat( newMedia ) );
+					// Reset lastMediaPassed and set it with urls from newMedia.
+					lastMediaPassed = newMedia.map( ( media ) => {
+						// Add everything up to '.fileType' to compare via .includes.
+						const cutOffIndex = media.url.lastIndexOf( '.' );
```

Comment:
If the user tries to upload two different images one named image-267x300.jpg and another named image.jpg, is there any chance a problem would occur?

Reply:
Great thought!

It looks like there is a problem when `some-name-X` is already uploaded, and then `some-name` is uploaded as a not-last member of a group upload.  But your idea mentioned below regarding comparing by ID if it exists seems to fix this case.